README.md

[comment]: # (Start Badges)

[![Build Status](https://travis-ci.org/47deg/github4s.svg?branch=master)](https://travis-ci.org/47deg/github4s) [![Join the chat at https://gitter.im/47deg/github4s](https://badges.gitter.im/47deg/github4s.svg)](https://gitter.im/47deg/github4s?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge) [![codecov.io](http://codecov.io/github/47deg/github4s/coverage.svg?branch=master)](http://codecov.io/github/47deg/github4s?branch=master) [![Maven Central](https://img.shields.io/badge/maven%20central-0.19.0-green.svg)](https://oss.sonatype.org/#nexus-search;gav~com.47deg~github4s*) [![License](https://img.shields.io/badge/license-Apache%202-blue.svg)](https://raw.githubusercontent.com/47deg/github4s/master/LICENSE) [![Latest version](https://img.shields.io/badge/github4s-0.19.0-green.svg)](https://index.scala-lang.org/47deg/github4s) [![Scala.js](http://scala-js.org/assets/badges/scalajs-0.6.17.svg)](http://scala-js.org) [![GitHub Issues](https://img.shields.io/github/issues/47deg/github4s.svg)](https://github.com/47deg/github4s/issues)

[comment]: # (End Badges)

Github4s
=============

**Github4s** is a GitHub API wrapper written in Scala.

## Installation

Github4s is compatible with both Scala JVM and Scala.js. It supports 2.10, 2.11 and 2.12.

To get started with SBT, simply add the following to your build.sbt file.

For Scala.jvm:

[comment]: # (Start Replace)

```scala
libraryDependencies += "com.47deg" %% "github4s" % "0.19.0"
```

[comment]: # (End Replace)

For Scala.js:

[comment]: # (Start Replace)

```scala
libraryDependencies += "com.47deg" %%% "github4s" % "0.19.0"
```

[comment]: # (End Replace)

## Github4s in the wild

If you wish to add your library here please consider a PR to include it in the list below.

★ | ★ | ★
--- | --- | ---
![scala-exercises](https://www.scala-exercises.org/assets/images/navbar_brand.svg) | [**scala-exercises**](https://www.scala-exercises.org/) | Scala Exercises is an Open Source project for learning different technologies based in the Scala Programming Language.
| | [**dashing**](https://github.com/benfradet/dashing) | Dashing is a collection of dashboards to monitor the health of an open source organization.

[comment]: # (Start Copyright)
# Copyright

Github4s is designed and developed by 47 Degrees

Copyright (C) 2016-2018 47 Degrees. <http://47deg.com>

[comment]: # (End Copyright)

CHANGELOG.md
# Changelog

## 10/04/2018 - Version 0.19.0

Release changes:

* Bump sbt-org-policies to 0.9.4 ([#223](https://github.com/47deg/github4s/pull/223))


## 09/04/2018 - Version 0.18.8

Release changes:

* Add dashing as an application using github4s ([#216](https://github.com/47deg/github4s/pull/216))
* Add listCollaborators method ([#218](https://github.com/47deg/github4s/pull/218))
* add get tree api implementation ([#217](https://github.com/47deg/github4s/pull/217))
* Release 0.18.7 ([#219](https://github.com/47deg/github4s/pull/219))
* Add cla-bot to the Github4s in the wild section ([#220](https://github.com/47deg/github4s/pull/220))
* Release 0.18.8 ([#221](https://github.com/47deg/github4s/pull/221))


## 08/20/2018 - Version 0.18.7

Release changes:

* Add dashing as an application using github4s ([#216](https://github.com/47deg/github4s/pull/216))
* Add listCollaborators method ([#218](https://github.com/47deg/github4s/pull/218))
* add get tree api implementation ([#217](https://github.com/47deg/github4s/pull/217))
* Release 0.18.7 ([#219](https://github.com/47deg/github4s/pull/219))


## 07/09/2018 - Version 0.18.6

Release changes:

* Abstract away from cats-effect IO ([#214](https://github.com/47deg/github4s/pull/214))
* Release 0.18.6 ([#215](https://github.com/47deg/github4s/pull/215))


## 06/24/2018 - Version 0.18.5

Release changes:

* List labels endpoint ([#206](https://github.com/47deg/github4s/pull/206))
* Bump Scala version to 2.12.6 in travis.yml ([#208](https://github.com/47deg/github4s/pull/208))
* Fix cats-effect JS tests flakiness ([#209](https://github.com/47deg/github4s/pull/209))
* Add support for the add labels endpoint ([#207](https://github.com/47deg/github4s/pull/207))
* Add support for the remove label endpoint ([#210](https://github.com/47deg/github4s/pull/210))
* Add support for the get pull request endpoint ([#211](https://github.com/47deg/github4s/pull/211))
* Release 0.18.5 ([#212](https://github.com/47deg/github4s/pull/212))


## 04/10/2018 - Version 0.18.4

Release changes:

* Keep circe decoders dry ([#202](https://github.com/47deg/github4s/pull/202))
* Change param so we are filtering by label ([#204](https://github.com/47deg/github4s/pull/204))
* Release 0.18.4 ([#205](https://github.com/47deg/github4s/pull/205))


## 03/11/2018 - Version 0.18.3

Release changes:

* # Make `body: Option[String]` ([#200](https://github.com/47deg/github4s/pull/200))
* Release 0.18.3 ([#201](https://github.com/47deg/github4s/pull/201))


## 03/07/2018 - Version 0.18.2

Release changes:

* Add pagination support for PullRequest ops ([#198](https://github.com/47deg/github4s/pull/198))
* Releases 0.18.2 ([#199](https://github.com/47deg/github4s/pull/199))


## 02/14/2018 - Version 0.18.1

Release changes:

* Fix Capture instances for IO and Future ([#195](https://github.com/47deg/github4s/pull/195))
* Release 0.18.1 ([#196](https://github.com/47deg/github4s/pull/196))


## 01/23/2018 - Version 0.18.0

Release changes:

* Bump ruby version in travis to fix travis build ([#182](https://github.com/47deg/github4s/pull/182))
* Make status' id a Long ([#183](https://github.com/47deg/github4s/pull/183))
* Make PullRequestFile#patch an Option[String] ([#181](https://github.com/47deg/github4s/pull/181))
* Fixes Travis file ([#190](https://github.com/47deg/github4s/pull/190))
* Bump sbt-org-policies to 0.8.22 to benefit from cats 1.0.1 and circe 0.9.1 ([#192](https://github.com/47deg/github4s/pull/192))
* Get a single issue ([#191](https://github.com/47deg/github4s/pull/191))
* Fetch issue comments ([#188](https://github.com/47deg/github4s/pull/188))
* Release version 0.18.0 ([#193](https://github.com/47deg/github4s/pull/193))


## 11/08/2017 - Version 0.17.0

Release changes:

* Bumps sbt-org-policies in order to fix docs autopublishing issue ([#168](https://github.com/47deg/github4s/pull/168))
* Bump sbt-org-policies to 0.8.7 to benefit from cats 1.0.0-RC1 & co ([#176](https://github.com/47deg/github4s/pull/176))
* Bump Travis' Scala version to 2.12.4 ([#177](https://github.com/47deg/github4s/pull/177))


## 09/25/2017 - Version 0.16.0

Release changes:

* Remove integration.GHGistsSpec ([#151](https://github.com/47deg/github4s/pull/151))
* Make gh4s doc structure reflect gh doc structure ([#152](https://github.com/47deg/github4s/pull/152))
* cats-effect module ([#155](https://github.com/47deg/github4s/pull/155))
* List statuses now gives back a 404 for a non-existing ref instead of an empty list ([#160](https://github.com/47deg/github4s/pull/160))
* Support for starring-related operations ([#161](https://github.com/47deg/github4s/pull/161))
* Support for the list organization repositories endpoint ([#162](https://github.com/47deg/github4s/pull/162))
* Organization API ([#163](https://github.com/47deg/github4s/pull/163))
* Bump sbt-org-policies to 0.7.4 ([#164](https://github.com/47deg/github4s/pull/164))
* Release 0.16.0 ([#165](https://github.com/47deg/github4s/pull/165))
* Bump travis 2.12 version to 2.12.3 ([#166](https://github.com/47deg/github4s/pull/166))


## 05/23/2017 - Version 0.15.0

Release changes:

* Remove integration tests creating statuses ([#128](https://github.com/47deg/github4s/pull/128))
* Streamlined the getting started ([#126](https://github.com/47deg/github4s/pull/126))
* Support Comment API ([#127](https://github.com/47deg/github4s/pull/127))
* Refactor Algebras ([#130](https://github.com/47deg/github4s/pull/130))
* Missing Test and Docs ([#132](https://github.com/47deg/github4s/pull/132))
* Missing unit test ([#137](https://github.com/47deg/github4s/pull/137))
* Contributing guide ([#138](https://github.com/47deg/github4s/pull/138))
* Replace updateReference's force parameter type from Option[Boolean] to Boolean ([#141](https://github.com/47deg/github4s/pull/141))
* Upgrades tut bumping sbt-org-policies version to 0.5.0 ([#140](https://github.com/47deg/github4s/pull/140))

project/ProjectPlugin.scala
import com.typesafe.sbt.site.jekyll.JekyllPlugin.autoImport._
import microsites._
import microsites.MicrositesPlugin.autoImport._
import sbt.Keys._
import sbt._
import sbtorgpolicies.model._
import sbtorgpolicies.OrgPoliciesKeys.orgBadgeListSetting
import sbtorgpolicies.OrgPoliciesPlugin
import sbtorgpolicies.OrgPoliciesPlugin.autoImport._
import sbtorgpolicies.templates.badges._
import sbtorgpolicies.runnable.syntax._
import scoverage.ScoverageKeys
import scoverage.ScoverageKeys._
import tut.TutPlugin.autoImport._

object ProjectPlugin extends AutoPlugin {

  override def trigger: PluginTrigger = allRequirements

  override def requires: Plugins = OrgPoliciesPlugin

  object autoImport {

    lazy val micrositeSettings = Seq(
      micrositeName := "Github4s",
      micrositeDescription := "Github API wrapper written in Scala",
      micrositeBaseUrl := "github4s",
      micrositeDocumentationUrl := "/github4s/docs.html",
      micrositeGithubOwner := "47deg",
      micrositeGithubRepo := "github4s",
      micrositeAuthor := "Github4s contributors",
      micrositeOrganizationHomepage := "https://github.com/47deg/github4s/blob/master/AUTHORS.md",
      micrositeExtraMdFiles := Map(
        file("CHANGELOG.md") -> ExtraMdFileConfig(
          "changelog.md",
          "page",
          Map("title" -> "Changelog", "section" -> "changelog", "position" -> "2")
        )
      ),
      includeFilter in Jekyll := "*.html" | "*.css" | "*.png" | "*.jpg" | "*.gif" | "*.js" | "*.swf" | "*.md",
      scalacOptions in Tut ~= (_ filterNot Set("-Ywarn-unused-import", "-Xlint").contains)
    )

    lazy val testSettings = Seq(
      fork in Test := false
    )

    lazy val commonCrossDeps = Seq(
      %%("cats-core"),
      %%("cats-free"),
      %%("simulacrum"),
      %%("circe-core"),
      %%("circe-generic"),
      %%("circe-parser"),
      %%("base64"),
      %%("frees-core"),
      %%("scalamockScalatest") % "test",
      %%("scalatest")          % "test"
    )

    lazy val standardCommonDeps = Seq(
      libraryDependencies += compilerPlugin(%%("scalameta-paradise") cross CrossVersion.full)
    )

    lazy val jvmDeps = Seq(
      libraryDependencies ++= Seq(
        %%("scalaj"),
        "org.mock-server" % "mockserver-netty" % "3.10.4" % "test" excludeAll ExclusionRule(
          "com.twitter")
      )
    )

    lazy val jsDeps: Def.Setting[Seq[ModuleID]] = libraryDependencies += %%%("roshttp")

    lazy val docsDependencies: Def.Setting[Seq[ModuleID]] = libraryDependencies += %%("scalatest")

    lazy val scalazDependencies: Def.Setting[Seq[ModuleID]] =
      libraryDependencies += %%("scalaz-concurrent")

    lazy val catsEffectDependencies: Seq[ModuleID] =
      Seq(
        %%("cats-effect"),
        %%("scalatest") % "test"
      )

    def toCompileTestList(sequence: Seq[ProjectReference]): List[String] = sequence.toList.map {
      p =>
        val project: String = p.asInstanceOf[LocalProject].project
        s"$project/test"
    }
  }

  override def projectSettings: Seq[Def.Setting[_]] =
    Seq(
      name := "github4s",
      orgProjectName := "Github4s",
      description := "Github API wrapper written in Scala",
      startYear := Option(2016),
      resolvers += Resolver.sonatypeRepo("snapshots"),
      scalaVersion := scalac.`2.12`,
      crossScalaVersions := "2.10.6" :: scalac.crossScalaVersions,
      scalacOptions ++= (scalaBinaryVersion.value match {
        case "2.10" => Seq("-Xdivergence211")
        case _      => Nil
      }),
      scalacOptions ~= (_ filterNot Set("-Xlint").contains),
      orgGithubTokenSetting := "GITHUB4S_ACCESS_TOKEN",
      orgBadgeListSetting := List(
        TravisBadge.apply(_),
        GitterBadge.apply(_),
        CodecovBadge.apply(_),
        MavenCentralBadge.apply(_),
        LicenseBadge.apply(_),
        ScalaLangBadge.apply(_),
        ScalaJSBadge.apply(_),
        GitHubIssuesBadge.apply(_)
      ),
      orgSupportedScalaJSVersion := Some("0.6.21"),
      orgScriptTaskListSetting ++= List(
        (ScoverageKeys.coverageAggregate in Test).asRunnableItemFull,
        "docs/tut".asRunnableItem
      ),
      coverageExcludedPackages := "<empty>;github4s\\.scalaz\\..*",
      // This is necessary to prevent packaging the BuildInfo with
      // sensible information like the Github token. Do not remove.
      mappings in (Compile, packageBin) ~= { (ms: Seq[(File, String)]) =>
        ms filter {
          case (_, toPath) =>
            !toPath.startsWith("github4s/BuildInfo")
        }
      }
    ) ++ shellPromptSettings
}
